### DISCUSSION TITLE
D3Q19 mwrt

### DISCUSSION THREAD

**Discussion by NickZhaooo** [CFD-GO/TCLB#407] [Q&A] [answered]
In source code, is there any functions or modules for wmrt model, especially for d3q19?

**Comment by llaniewski** [comment]
@TravisMitchell Can you point to a WMRT generation example?

**Reply by NickZhaooo** [reply]
Of coursre. When we process the collision step with WMRT model, O = -M*S*M*(f-f_eq) is implemented. M is the collision matrix in D3Q19, which consists of a series equations, such as, 
m0 = |ci|^0
m1 = cx
m2 = cy
m3 = cz
...
Above equaitons are used to generate the matrix M.
I searched source codes, and I can't find the codes about M.

**Comment by llaniewski** [comment, marked as answer]
Generally in https://github.com/CFD-GO/TCLB/blob/master/src/lib/feq.R you have everything you need. You can generate an object, which carries all the information for MRT:
```R
library(polyAlgebra)
source("src/lib/feq.R")
source("src/lib/lattice.R")

# Velocity set
U = d3q19

# symbolic variables
f = PV("f",1:nrow(U)-1)
R = PV("R",1:nrow(U)-1)

# object that has all the MRT things
EQ = MRT_eq(U)

# example collision
C( R, f %*% EQ$mat )
C( PV("rho","Jx","Jy","Jz"), R[1:4] )
C( R, R + (EQ$Req - R)*PV("omega"))
C( f, R %*% solve(EQ$mat) )
```

To do WMRT (weight MRT)
```R
# get rid of the hardcoded MRT matrix from literature
attr(U, "MAT") = NULL

# generate WMRT matrix
M = WMRT_mat(U)

# rest is the same, generate an object and do collision
EQ = MRT_eq(U, mat=M)

C( R, f %*% EQ$mat )
C( PV("rho","Jx","Jy","Jz"), R[1:4] )
C( R, R + (EQ$Req - R)*PV("omega"))
C( f, R %*% solve(EQ$mat) )
```

**Reply by NickZhaooo** [reply]
Thanks, I got it. GOOD JOBS.
